crypto/cms/cms_env.c

```diff
@@ -1031,26 +1073,57 @@ BIO *cms_EnvelopedData_init_bio(CMS_ContentInfo *cms)
  */
 int cms_pkey_get_ri_type(EVP_PKEY *pk)
 {
+#if 0                            /* This should just be removed */
     if (pk->ameth && pk->ameth->pkey_ctrl) {
         int i, r;
         i = pk->ameth->pkey_ctrl(pk, ASN1_PKEY_CTRL_CMS_RI_TYPE, 0, &r);
         if (i > 0)
             return r;
     }
     return CMS_RECIPINFO_TRANS;
+#else
+    /*
+     * We try to figure out what the key type supports by doing one init
+     * after the other.
+     */
+    EVP_PKEY_CTX *pctx = EVP_PKEY_CTX_new(pk, NULL);
+    int r;
+
+    ERR_set_mark();
+    if (EVP_PKEY_derive_init(pctx) > 0)
+        r = CMS_RECIPINFO_AGREE;
+    else if (EVP_PKEY_encrypt_init(pctx) > 0)
+        r = CMS_RECIPINFO_TRANS;
```

Comment:
I suggest at least set this branch first.

Reply:
So it seems GOST case at least would be fixed by checking for encrypt first and derive later. Can we rationalize this and leave a comment explaining why a specific order is preferable?

Reply:
I have nothing against switching them around, but what will happen then with other implementations that have stronger derive than encrypt / decrypt support?  It's basically the same problem as @beldmit describes, but the other way around.

Although, I don't know CMS very well, so it's possible encrypt should always be used first and derive as fallback...  if someone tells me that's always the case, I'll happily switch.

Either way, I could use some help formulating a comment to explain the order we choose.

Reply:
Well, the problem is deeper.
We potentially have more than one supported format both for encryption and description. In the case of decryption, I think it's more reasonable to delegate the support to an engine/provider. For encryption, we currently don't have appropriate control function call to choose the preferred way and don't have any implementation that requires such a control.

Reply:
> In the case of decryption, I think it's more reasonable to delegate the support to an engine/provider.

You mean to still ask the engine/provider the direct question instead of probing?  I'm *not* too keen on insert CMS specific hackery in providers, this PR is an effort to get away from that...  but will have to give this some though.  A capability probe maybe?

If I understood a little better what possibilities there are and on what conditions, that might help.

Reply:
I'm not aware of all the possible options either. RFC 4490 describes both variants for GOST, but in practice, we met only KeyTransport (partially because KeyAgreement was not implemented in OpenSSL at all). Do I understand correctly that RSA, DSA, and EC provide only one variant?

BTW, it's worth mentioning that capability of key encryption/derivation does not say us anything about CMS support.

Reply:
I'm not sure what "CMS support" should be needed on the provider side.  After all, key derivation with the same algorithms is done the same way regardless, and so is encryption, right?  After all, looking at our EVP_PKEY_ASN1_METHODs for DH, DSA, EC and RSA, their ASN1_PKEY_CTRL_CMS_RI_TYPE responses are, respectively, CMS_RECIPINFO_AGREE, CMS_RECIPINFO_NONE, CMS_RECIPINFO_AGREE and CMS_RECIPINFO_TRANS.  To me, those look like the absolutely obvious answer, which could just as well have been probed by looking if the corresponding EVP_PKEY_METHOD has a non-NULL `encrypt_init` or `derive_init`...

Looking at the old code, this function defaults to KeyTransport if there's no response on ASN1_PKEY_CTRL_CMS_RI_TYPE, so I guess that's how this new code differs, and since there a number of functions in gost_ameth.c that support both  derivation and encryption, and that's likely to be the case in a provider as well, then the sudden change of primary choice may come as a bit of a surprise, I get that.

So I guess that's the hard question to ask, then...  in case of ambiguity, who should have the power to decide which method should take priority?

Reply:
I'll ponder this over-night...  or I guess that'll be in the morning, 'cause I'm not very awake for the moment.

Reply:
In the interest of expedience, I've simply switched them around so KeyTransport is probed first, and added a TODO comment that we should investigate that further.  That will allow us to move forward with this PR, and perhaps let our brains ruminate on the problem for some time and revisit it later.